ion-schema/src/nfa.rs

```diff
@@ -4,135 +4,219 @@ use crate::types::TypeValidator;
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct FinalState {
+    state_id: StateId,
+    min: usize,
+    max: usize,
+}
+
+impl FinalState {
+    pub fn new(state_id: StateId, min: usize, max: usize) -> Self {
+        Self { state_id, min, max }
+    }
+}
+
+/// Represents an NFA that will eb used by the ordered_elements` constraint in order to validate an Ion Value
 #[derive(Default, Debug, Clone, PartialEq)]
 pub struct Nfa {
-    pub(crate) states: Vec<State>,
-    pub(crate) transitions: HashMap<usize, HashSet<Transition>>,
+    pub(crate) transitions: HashMap<StateId, HashSet<Transition>>, // represents transitions between states
+    pub(crate) final_states: HashSet<FinalState>, // represents all the final states for NFA
 }
 
-#[derive(Debug, Clone)]
+impl Nfa {
+    pub fn get_final_states(&self) -> HashSet<FinalState> {
+        self.final_states.to_owned()
+    }
+
+    pub fn get_transitions(&self, state_id: StateId) -> HashSet<Transition> {
+        self.transitions
+            .get(&state_id)
+            .map(|s| s.to_owned())
+            .unwrap_or(HashSet::new())
+    }
+}
+
+/// A context that will be used for a single transition
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
 pub struct NfaRun {
-    pub(crate) visits: HashMap<usize, usize>,
-    pub(crate) nfa: Rc<Nfa>,
+    state_id: StateId,
+    state_visits: usize,
 }
 
 impl NfaRun {
+    pub fn new(current_state_id: StateId, visits_to_current_state: usize) -> Self {
+        Self {
+            state_id: current_state_id,
+            state_visits: visits_to_current_state,
+        }
+    }
+}
+
+/// This is a context which will be used while validating an Ion value for `ordered_elements` constraint using its NFA
+#[derive(Debug, Clone)]
+pub struct NfaEvaluation {
+    pub(crate) visits: HashSet<NfaRun>,
+    pub(crate) nfa: Rc<Nfa>,
+}
+
+impl NfaEvaluation {
     pub fn new(nfa: Rc<Nfa>) -> Self {
         Self {
             visits: {
-                let mut visits = HashMap::new();
-                visits.insert(0, 1);
+                let mut visits = HashSet::new();
+                if !nfa.get_final_states().is_empty() {
+                    visits.insert(NfaRun::new(0, 0));
+                }
                 visits
             },
             nfa,
         }
     }
 
-    pub fn has_final_state(&self) -> bool {
-        self.visits
-            .iter()
-            .any(|(id, _)| self.nfa.states[*id] == State::Final)
+    pub fn has_final_state(&self, type_store: &TypeStore) -> bool {
+        if self.nfa.get_final_states().is_empty() && self.visits.is_empty() {
+            return true;
+        }
+
+        self.visits.iter().any(|nfa_run| {
+            self.nfa.get_final_states().iter().any(|fs| {
+                fs.state_id == nfa_run.state_id
+                    && nfa_run.state_visits >= fs.min
+                    && nfa_run.state_visits <= fs.max
```

Comment:
It is validated at each transition but when an invalid Ion value is passed through transitions its possible that it reaches final state but with incorrect count. Meaning if the invalid Ion value reaches final state but doesn't satisfy minimum occurrence constraint then it would lead to an incorrect visit count. 

Example from `ion-schema-tests`:
```
// for the following schema:
type::{
  name: ordered_elements_occurs_1,
  ordered_elements: [
    { type: int, occurs: 2 }, // S0
    { type: string, occurs: range::[0, 3] }, // S1
    { type: symbol, occurs: range::[0, max] }, // S2
  ],
}

// Invalid Ion value:
(1)
(1 "a" w)

// nfa_run results fro `(1)`
1 --> {NfaRun { state_id: 0, state_visits: 1 }}

// nfa_run results fro `(1 "a" w)`
1 --> {NfaRun { state_id: 0, state_visits: 1 }}
"a" --> {NfaRun { state_id: 0, state_visits: 1 }}
w --> {NfaRun { state_id: 0, state_visits: 1 }}
```

Reply:
Alternate would be to remove final state from next set of `NfaRun`s when a minimum occurrence isn't met.